src/game/shadow.c

```diff
@@ -244,93 +192,152 @@ Gfx *create_shadow_below_xyz(Vec3f pos, s16 shadowScale, u8 shadowSolidity, s8 s
         return NULL;
     }
 
-    register f32 x = pos[0];
-    register f32 y = pos[1];
-    register f32 z = pos[2];
+    // The floor underneath the object.
+    struct Surface *floor = NULL;
+    // The y-position of the floor (or water or lava) underneath the object.
+    f32 floorHeight = FLOOR_LOWER_LIMIT_MISC;
+    f32 x = pos[0];
+    f32 y = pos[1];
+    f32 z = pos[2];
     s8 isPlayer   = (obj == gMarioObject);
     s8 notHeldObj = (gCurGraphNodeHeldObject == NULL);
 
     // Attempt to use existing floors before finding a new one.
     if (notHeldObj && isPlayer && gMarioState->floor) {
-        // Object is player and has a referenced floor.
-        s->floor       = gMarioState->floor;
-        s->floorHeight = gMarioState->floorHeight;
+        // The object is Mario and has a referenced floor.
+        floor       = gMarioState->floor;
+        floorHeight = gMarioState->floorHeight;
     } else if (notHeldObj && (gCurGraphNodeObject != &gMirrorMario) && obj->oFloor) {
-        // Object is not player but has a referenced floor.
+        // The object is not Mario but has a referenced floor.
         //! Some objects only get their oFloor from bhv_init_room, which skips dynamic floors.
-        s->floor       = obj->oFloor;
-        s->floorHeight = obj->oFloorHeight;
+        floor       = obj->oFloor;
+        floorHeight = obj->oFloorHeight;
     } else {
-        // Object has no referenced floor, so find a new one.
+        // The object has no referenced floor, so find a new one.
         // gCollisionFlags |= COLLISION_FLAG_RETURN_FIRST;
```

Comment:
i'd like to see if this would be adequate for shadows, perfect shadow collision probably isn't worth the extra CPU per shadow (and sometimes there's a LOT of shadows)

Reply:
https://user-images.githubusercontent.com/79979276/147256224-5037b65c-4847-40ee-82da-7b0b5e328f61.mp4

___

So the top one is without the flag, the bottom is with the flag. With `COLLISION_FLAG_RETURN_FIRST` ,  `collision` time is roughly 2650-2800us on average, without the flag collision time is roughly 3150-3300. Which leaves me at a 500us save in this spot specifically, which is a pretty shadow heavy area. So far I haven't seen issues with the flag on, but 500us in a shadow-heavy is probably not significant enough to consider for 2.0.

@Mr-Wiseguy thoughts?

Reply:
https://user-images.githubusercontent.com/79979276/147257840-3aab3e2c-0e35-4ba7-8cab-23c4c04cb1a8.mp4

NEVERMIND, so with the flag this happens. Thats super weird though, it should be getting that floor >:( something seems to be up with our collision code if returning the first one is this bad. There's literally no floor besides this one and the death floor beneath the bullet bill.